src/psyclone/domain/gocean/transformations/kern_trans.py

```diff
@@ -0,0 +1,163 @@
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author R. W. Ford STFC Daresbury Lab
+
+'''Raise generic PSyIR representing a kernel-layer routine to
+PSyclone kernel-layer-specific PSyIR which uses specialised classes.
+
+'''
+from psyclone.domain.gocean.kernel import KernelMetadataSymbol
+from psyclone.psyGen import Transformation
+from psyclone.psyir.nodes import Schedule, Container
+from psyclone.psyir.transformations import TransformationError
+
+
+class KernTrans(Transformation):
+    '''Raise a generic PSyIR representation of a kernel-layer routine
+    to a PSyclone version with specialised domain-specific nodes and
+    symbols. This is currently limited to the specialisation of kernel
+    metadata.
+
+    '''
+    def __init__(self):
+        super().__init__()
+        self._metadata_name = None
+
+    def validate(self, node, options=None):
+        '''Validate the supplied PSyIR tree.
+
+        :param node: a PSyIR node that is the root of a PSyIR tree.
+        :type node: Union[:py:class:`psyclone.psyir.node.Routine`, \
+            :py:class:`psyclone.psyir.node.Container`]
+        :param options: a dictionary with options for transformations.
+        :type options: Optional[Dict[str: str]]
+
+        :raises TransformationError: if the metadata name has not been \
+            set or does not exist in the code.
+        :raises TransformationError: if the supplied node argument is \
+            not a Routine or a Container.
+        :raises TransformationError: if the supplied node argument has \
+            a parent.
+
+        '''
+        if not self._metadata_name:
+            raise TransformationError(
```

Comment:
Should this be an InternalError - is it up to an API-specific sub-class to set this?

Reply:
Oh, but I see that this is an API specific subclass. How come _metadata_name is a parameter - I must be missing something?

Reply:
This tells the transformation the name used to store the metadata.